src/components/ProviderSelector.tsx

```diff
@@ -0,0 +1,151 @@
+import { ApiPromise, WsProvider } from "@polkadot/api";
+import { hexToU8a } from "@polkadot/util";
+import { base58Encode } from "@polkadot/util-crypto";
+import { AlertCircle, Loader2, RefreshCw, Server } from "lucide-react";
+import { useEffect, useRef, useState } from "react";
+import { ProviderButton } from "./buttons/ProviderButton";
+
+const COLLATOR_RPC_URL = "ws://127.0.0.1:42069"; // TODO: replace with some mechanism like polkadot.js
+
+enum Status {
+  Connecting = 0,
+  Connected = 1,
+  Failed = 2,
+}
+
+type ProviderSelectorProps = {
+  providers: Map<string, object>;
+  setProviders: (providers: Map<string, object>) => void;
+  onSelectProvider: (provider: string) => void;
+  selectedProviders: Set<string>;
+};
+
+export function ProviderSelector({
+  providers,
+  setProviders,
+  onSelectProvider,
+  selectedProviders,
+}: ProviderSelectorProps) {
+  const [error, setError] = useState<string | null>(null);
+  const [status, setStatus] = useState(Status.Connecting);
+
+  const apiPromiseRef = useRef<ApiPromise | null>(null);
+
+  // A liveness check before populating (or maybe populating but disabled)
+  // would be great UX
+  const getStorageProviders = async () => {
+    setStatus(Status.Connecting);
+    setError(null);
+    setProviders(new Map());
+
+    // This should be parametrizable
+    const wsProvider = new WsProvider(COLLATOR_RPC_URL);
+    try {
+      console.log(`Connecting to ${COLLATOR_RPC_URL}`);
+      const polkaStorageApi = await ApiPromise.create({
+        provider: wsProvider,
+      });
+
+      const entries = await polkaStorageApi.query.storageProvider.storageProviders.entries();
+      for (const [key, value] of entries) {
+        const human = value.toHuman().info; // TODO: proper conversion
+        if ("peerId" in human) {
+          human.peerId = base58Encode(hexToU8a(human.peerId));
+        }
+        const newProviders = new Map(providers);
+        // debugger;
+        newProviders.set(key.args[0].toString(), human);
+        setProviders(newProviders);
+      }
+
+      console.log(`Connected to ${await polkaStorageApi.rpc.system.chain()}`);
+      apiPromiseRef.current = polkaStorageApi;
+      setStatus(Status.Connected);
+    } catch (error) {
+      if (error && error instanceof Error) {
+        setStatus(Status.Failed);
+        setError(error.message);
+      }
+    }
+  };
+
+  // biome-ignore lint/correctness/useExhaustiveDependencies: False positive, this is supposed to run ONCE
+  useEffect(() => {
+    getStorageProviders();
+  }, []);
```

Comment:
Not necessarily a false positive, this function changes to a different object every rerender. I'd say it should be wrapped in [`useCallback` ](https://react.dev/reference/react/useCallback)